### PR TITLE
victoriametrics: fix seriesByTag name tag mapping and quote-aware tag splitting

### PR DESCRIPTION
**Pull request by rassvetalov**
## Summary

Two fixes for `seriesByTag(...)` conversion to PromQL:

1. **VictoriaMetrics: map `name` tag to `__graphite__` instead of `__name__`.**
   VictoriaMetrics stores Graphite metric paths in the `__graphite__` label. The non-seriesByTag code path already uses `{__graphite__=...}` for plain targets. However, the seriesByTag branch used `strings.ReplaceAll` to rewrite `'name='` to `'__name__='`, which is incorrect for VM (selects the wrong label) and fragile (matches substrings like `'hostname='`). Replaced with a new `SeriesByTagToPromQLWithRenames` helper that renames tags at the parsed-map level.

2. **Quote-aware comma splitting in `SplitTagValues`.**
   `SplitTagValues` used `strings.Split(query, ",")` which incorrectly splits tag values that contain commas (e.g. `seriesByTag('desc=hello, world')`). Added `splitWithQuotes` that respects single/double-quoted regions. Also added a guard against panics on malformed (too-short) segments.

## Changes

- `zipper/protocols/prometheus/helpers/helpers.go`:
  - Add `splitWithQuotes(s, delimiter)` — quote-aware string splitter.
  - `SplitTagValues` now uses `splitWithQuotes` and skips segments shorter than 2 chars.
  - Add `SeriesByTagToPromQLWithRenames(step, target, tagRenames)` — applies a tag rename map after parsing, before building PromQL. The original `SeriesByTagToPromQL` is preserved as a zero-rename wrapper (fully backward-compatible).
- `zipper/protocols/victoriametrics/fetch.go`:
  - Replace `strings.ReplaceAll(target.name, "'name=", "'__name__=")` + `SeriesByTagToPromQL` with `SeriesByTagToPromQLWithRenames(..., {"name": "__graphite__"})`.
- `zipper/protocols/prometheus/helpers/helpers_test.go`:
  - Add `TestSplitTagValues` (simple, comma-in-value, empty, single tag).
  - Add `TestSeriesByTagToPromQLWithRenames` (name to __graphite__ rename, hostname unaffected, nil renames).

## Backward compatibility

- `SeriesByTagToPromQL` signature and behavior unchanged. All existing callers (Prometheus protocol) are unaffected.
- Only the VictoriaMetrics Fetch path changes behavior: `name` tag now maps to `__graphite__` instead of `__name__`.

## Test plan

- [ ] Existing `helpers_test.go` tests pass.
- [ ] New `TestSplitTagValues` covers comma-in-value edge case.
- [ ] New `TestSeriesByTagToPromQLWithRenames` covers rename logic and no-false-positive on similar tag names (e.g. `hostname`).
- [ ] Manual test with VictoriaMetrics + Graphite ingestion: `seriesByTag('name=servers.web.cpu')` should produce `{__graphite__="servers.web.cpu"}`.

### CODE REVIEW

File: zipper/protocols/prometheus/helpers/helpers.go [hunk 1]
```diff
@@ -177,12 +177,37 @@ func PromethizeTagValue(tagValue string) (string, types.Tag) {
...
+		switch {
+		case c == quoteChar:
+			quoteChar = 0
+		case quoteChar == 0 && (c == '\'' || c == '"'):
+			quoteChar = c
+		case quoteChar == 0 && c == delimiter:
+			result = append(result, s[start:i])
+			start = i + 1
+		}
+	}
```
[the remark this answers is not in the document]
Comment (deniszh) [reply zipper/protocols/prometheus/helpers/helpers.go:197]: Don't think we care about backslash-escaped quotes, but can be added.

File: zipper/protocols/prometheus/helpers/helpers.go [hunk 2]
```diff
@@ -210,13 +235,34 @@ func PromMetricToGraphite(metric map[string]string) string {
...
 	tagsString := target[len("seriesByTag(") : len(target)-1]
 	tvs := SplitTagValues(tagsString)
+
+	// Apply tag renames (e.g. "name" -> "__graphite__" for VictoriaMetrics).
+	for oldName, newName := range tagRenames {
+		if v, ok := tvs[oldName]; ok {
+			delete(tvs, oldName)
+			tvs[newName] = v
+		}
+	}
```
[the remark this answers is not in the document]
Comment (deniszh) [reply zipper/protocols/prometheus/helpers/helpers.go:267]: Also, don't think that can be the case. In theory maybe, but VM should be severily broken to return different name and graphite name tags.

File: zipper/protocols/prometheus/helpers/helpers_test.go [hunk 3]
```diff
@@ -250,3 +250,91 @@ func TestPromethizeTagValue(t *testing.T) {
...
+			wantPromQL: `{__graphite__="servers.web.cpu", env="prod"}`,
+		},
+		{
+			name:       "VM: hostname not affected by name rename",
+			step:       "30",
+			target:     "seriesByTag('hostname=web1','env=prod')",
+			renames:    map[string]string{"name": "__graphite__"},
+			wantStep:   "30",
+			wantPromQL: `{hostname="web1", env="prod"}`,
+		},
```
[the remark this answers is not in the document]
Comment (deniszh) [reply zipper/protocols/prometheus/helpers/helpers_test.go:345]: This looks like valid, @rassvetalov ? WDYT?

File: zipper/protocols/prometheus/helpers/helpers_test.go [hunk 4]
```diff
@@ -250,3 +250,91 @@ func TestPromethizeTagValue(t *testing.T) {
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := SplitTagValues(tt.query)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (deniszh) [reply zipper/protocols/prometheus/helpers/helpers_test.go:319]: Kinda nitpicking 💀 
But author can decide.

File: zipper/protocols/prometheus/helpers/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rassvetalov) [reply zipper/protocols/prometheus/helpers/helpers.go:197]: Thanks for the suggestion. As @deniszh noted, backslash-escaped quotes are not used in Graphite seriesByTag syntax, so this case doesn't apply in practice. The function now documents its behavior with mismatched quotes (the rest of the string is treated as quoted).

I've added edge-case tests for trailing delimiters, whitespace, and double-quoted values.

File: zipper/protocols/prometheus/helpers/helpers_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rassvetalov) [reply zipper/protocols/prometheus/helpers/helpers_test.go:319]: Added edge-case tests: trailing comma (empty segment skipped by the len<2 guard), whitespace around tags, and double-quoted values.

File: zipper/protocols/prometheus/helpers/helpers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rassvetalov) [reply zipper/protocols/prometheus/helpers/helpers.go:267]: Good catch. Documented this as last-write-wins in the function comment and added a test case ("rename collision: existing tag overwritten") that explicitly validates this behavior.

File: zipper/protocols/prometheus/helpers/helpers_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rassvetalov) [reply zipper/protocols/prometheus/helpers/helpers_test.go:345]: Fixed. Tag names are now sorted before building the PromQL string (using sort.Strings), which matches the pattern used in PromMetricToGraphite. Tests updated with expected sorted order, and verified stable across 5 consecutive runs with -count=5.

**Review by deniszh — APPROVED** [review]

**Comment by deniszh** [comment]
LGTM, merging this!
Thanks a lot for your contribution, @rassvetalov !
